src/data/astarte_provider.rs

```diff
@@ -0,0 +1,242 @@
+/*
+ * This file is part of Astarte.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::collections::HashMap;
+use std::io::Error;
+use std::sync::Arc;
+
+#[cfg(not(test))]
+use astarte_sdk::AstarteSdk;
+use async_trait::async_trait;
+use log::warn;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::RwLock;
+use tonic::Status;
+use uuid::Uuid;
+
+use crate::astarte_message_hub::AstarteNode;
+use crate::data::astarte::AstarteSubscriber;
+#[cfg(test)]
+use crate::data::mock_astarte::MockAstarteSdk as AstarteSdk;
+use crate::proto_message_hub;
+use crate::proto_message_hub::AstarteMessage;
+
+pub struct Astarte {
+    pub device_sdk: AstarteSdk,
+    subscribers: Arc<RwLock<HashMap<Uuid, Subscriber>>>,
+}
+
+struct Subscriber {
+    introspection: Vec<proto_message_hub::Interface>,
+    sender: Sender<Result<AstarteMessage, Status>>,
+}
+
+#[async_trait]
+impl AstarteSubscriber for Astarte {
+    async fn subscribe(
+        &self,
+        astarte_node: &AstarteNode,
+    ) -> Result<Receiver<Result<AstarteMessage, Status>>, Error> {
+        let (tx, rx) = channel(2);
```

Comment:
Why this channel has size 2?

Reply:
I think that 2 doesn't make any sense, maybe is better increment to 32 as reported into `tokio documentation` as default, after that when we will make tests on real device, we can will decide if increase or decrease. What do you think?

Reply:
Ok!